File: cli/scripts/coevolution.py
"""Compute mutual information co-evolution analysis from aligned FASTA."""
from __future__ import annotations
import io
import math
from collections import Counter

import numpy as np
from Bio import AlignIO

from scripts.conservation_metrics import GAP_CHARS, compute_gap_zscores, gap_fraction


def _consensus(col: list[str]) -> str:
    """Most common non-gap residue in a column, or '?' if all gaps."""
    residues = [a for a in col if a not in GAP_CHARS]
    return Counter(residues).most_common(1)[0][0] if residues else "?"


def mutual_information(col_i: list[str], col_j: list[str]) -> float:
    """Mutual information (bits) between two alignment columns, gaps excluded."""
    pairs = [
        (a, b) for a, b in zip(col_i, col_j)
        if a not in GAP_CHARS and b not in GAP_CHARS
    ]
    if not pairs:
        return 0.0
    n = len(pairs)
    freq_i = Counter(a for a, _ in pairs)
    freq_j = Counter(b for _, b in pairs)
    freq_ij = Counter(pairs)
    mi = 0.0
    for (a, b), cnt in freq_ij.items():
        p_ij = cnt / n
        p_i = freq_i[a] / n
        p_j = freq_j[b] / n
        mi += p_ij * math.log2(p_ij / (p_i * p_j))
    return float(max(0.0, mi))


def apply_apc(mi: np.ndarray) -> np.ndarray:
    """
    Apply Average Product Correction (Dunn et al. 2008) to a symmetric MI matrix.
    Row means and global mean are computed excluding the diagonal (self-MI is undefined).
    Uses numpy masked arrays to safely exclude diagonal without reshape assumptions.
    """
    n = mi.shape[0]
    # Mask diagonal: True = excluded from calculations
    masked = np.ma.array(mi, mask=np.eye(n, dtype=bool))
    mean_all = float(masked.mean())
    if mean_all == 0.0:
        return mi.copy()
    # Per-row off-diagonal mean: masked.mean(axis=1) ignores the masked diagonal
    mean_i = np.array(masked.mean(axis=1))  # shape (n,)
    apc = np.outer(mean_i, mean_i) / mean_all
    return np.maximum(0.0, mi - apc)


def compute_coevolution(
    aligned_fasta: str,
    ref_accession: str,
    gap_zscore_k: float = 1.5,
    top_n: int = 50,
    cluster: str = "UniRef50",  # reserved for future use; not currently included in output
) -> dict:
    """
    Compute MI co-evolution matrix with APC correction.
    Returns dict conforming to coevolution.schema.json v1.
    Positions with gap_zscore > gap_zscore_k are excluded (sparse positions array).
    """
    alignment = AlignIO.read(io.StringIO(aligned_fasta), "fasta")
    n_seqs = len(alignment)
    n_cols = alignment.get_alignment_length()

    # Find reference row to get residue numbering
    ref_row = next(
        (rec for rec in alignment if ref_accession in rec.id), alignment[0]
    )
    ref_seq = str(ref_row.seq)

    all_columns = [[str(alignment[i, col]) for i in range(n_seqs)] for col in range(n_cols)]
    all_gap_fracs = [gap_fraction(col) for col in all_columns]
    gap_zscores = compute_gap_zscores(all_gap_fracs)

    # Select non-gap reference positions that pass gap filter
    included: list[tuple[int, int]] = []  # (ref_residue_number, col_idx)
    ref_pos = 0
    for col_idx, ref_aa in enumerate(ref_seq):
        if ref_aa in GAP_CHARS:
            continue
        ref_pos += 1
        if gap_zscores[col_idx] <= gap_zscore_k:
            included.append((ref_pos, col_idx))

    positions = [rp for rp, _ in included]
    n_pos = len(positions)
    if n_pos == 0:
        raise ValueError(
            f"No positions passed the gap filter (gap_zscore_k={gap_zscore_k}). "
            "Try increasing gap_zscore_k."
        )

    # Build MI matrix over included columns
    included_cols = [all_columns[ci] for _, ci in included]
    mi_matrix = np.zeros((n_pos, n_pos))
    for i in range(n_pos):
        for j in range(i + 1, n_pos):
            val = mutual_information(included_cols[i], included_cols[j])
            mi_matrix[i, j] = mi_matrix[j, i] = val

    apc_matrix = apply_apc(mi_matrix)

    # Build top pairs
    pairs = []
    for i in range(n_pos):
        for j in range(i + 1, n_pos):
            if mi_matrix[i, j] > 0:
                pairs.append({
                    "position_i": positions[i],
                    "position_j": positions[j],
                    "aa_i": _consensus(included_cols[i]),
                    "aa_j": _consensus(included_cols[j]),
                    "mi_score": round(float(mi_matrix[i, j]), 6),
                    "mi_score_apc": round(float(apc_matrix[i, j]), 6),
                })
    pairs.sort(key=lambda p: p["mi_score_apc"], reverse=True)
    top_pairs = pairs[:top_n]

    return {
        "schema_version": "1",
        "accession": ref_accession,
        "method": "mutual_information_apc",
        "n_positions": n_pos,
        "positions": positions,
        "matrix": apc_matrix.tolist(),
        "top_pairs": top_pairs,
    }

File: webapp/src/analysis/coevolution.js
/**
 * Co-evolution: mutual information + APC correction.
 * Mirrors cli/scripts/coevolution.py exactly.
 */

import { GAP_CHARS, computeGapZscores, gapFraction } from "./conservation.js";

export function mutualInformation(colI, colJ) {
  const pairs = colI.map((a, k) => [a, colJ[k]])
    .filter(([a, b]) => !GAP_CHARS.has(a) && !GAP_CHARS.has(b));
  if (pairs.length === 0) return 0;
  const n = pairs.length;
  const freqI = {}, freqJ = {}, freqIJ = {};
  for (const [a, b] of pairs) {
    freqI[a] = (freqI[a] || 0) + 1;
    freqJ[b] = (freqJ[b] || 0) + 1;
    const key = `${a}|${b}`;
    freqIJ[key] = (freqIJ[key] || 0) + 1;
  }
  let mi = 0;
  for (const [key, cnt] of Object.entries(freqIJ)) {
    const [a, b] = key.split("|");
    const pIJ = cnt / n;
    const pI = freqI[a] / n;
    const pJ = freqJ[b] / n;
    mi += pIJ * Math.log2(pIJ / (pI * pJ));
  }
  return Math.max(0, mi);
}

export function applyApc(mi) {
  // mi is a 2D array (n×n), symmetric, zero diagonal
  const n = mi.length;
  // Compute row means and global mean excluding diagonal
  const rowMeans = mi.map((row, i) => {
    const offDiag = row.filter((_, j) => j !== i);
    return offDiag.reduce((a, b) => a + b, 0) / offDiag.length;
  });
  const allOffDiag = mi.flatMap((row, i) => row.filter((_, j) => j !== i));
  const globalMean = allOffDiag.reduce((a, b) => a + b, 0) / allOffDiag.length;
  if (globalMean === 0) return mi.map(row => [...row]);

  return mi.map((row, i) =>
    row.map((val, j) => {
      const apc = (rowMeans[i] * rowMeans[j]) / globalMean;
      return Math.max(0, val - apc);
    })
  );
}

/**
 * Compute co-evolution matrix from alignment.
 * @param {Array<{id: string, seq: string}>} alignment
 * @param {string} refAccession
 * @param {number} gapZscoreK
 * @param {number} topN
 * @returns {Object} conforming to coevolution.schema.json v1
 */
export function computeCoevolution(alignment, refAccession, gapZscoreK = 1.5, topN = 50) {
  const nSeqs = alignment.length;
  const nCols = alignment[0].seq.length;
  const refRow = alignment.find(r => r.id.includes(refAccession)) || alignment[0];

  const columns = Array.from({ length: nCols }, (_, ci) => alignment.map(r => r.seq[ci]));
  const allGapFracs = columns.map(gapFraction);
  const gapZscores = computeGapZscores(allGapFracs);

  // Included positions: non-gap in ref + gap_zscore <= k
  const included = [];
  let refPos = 0;
  for (let ci = 0; ci < nCols; ci++) {
    if (GAP_CHARS.has(refRow.seq[ci])) continue;
    refPos++;
    if (gapZscores[ci] <= gapZscoreK) included.push({ refPos, ci });
  }

  const positions = included.map(p => p.refPos);
  const nPos = positions.length;
  const includedCols = included.map(p => columns[p.ci]);

  // Build MI matrix
  const miMatrix = Array.from({ length: nPos }, () => Array(nPos).fill(0));
  for (let i = 0; i < nPos; i++) {
    for (let j = i + 1; j < nPos; j++) {
      const val = mutualInformation(includedCols[i], includedCols[j]);
      miMatrix[i][j] = miMatrix[j][i] = val;
    }
  }
  const apcMatrix = applyApc(miMatrix);

  // Top pairs
  const pairs = [];
  for (let i = 0; i < nPos; i++) {
    for (let j = i + 1; j < nPos; j++) {
      if (miMatrix[i][j] > 0) {
        const consensus = col => {
          const r = col.filter(a => !GAP_CHARS.has(a));
          if (!r.length) return "?";
          const f = {};
          r.forEach(a => f[a] = (f[a] || 0) + 1);
          return Object.entries(f).sort((a, b) => b[1] - a[1])[0][0];
        };
        pairs.push({
          position_i: positions[i],
          position_j: positions[j],
          aa_i: consensus(includedCols[i]),
          aa_j: consensus(includedCols[j]),
          mi_score: +miMatrix[i][j].toFixed(6),
          mi_score_apc: +apcMatrix[i][j].toFixed(6),
        });
      }
    }
  }
  pairs.sort((a, b) => b.mi_score_apc - a.mi_score_apc);

  return {
    schema_version: "1",
    accession: refAccession,
    method: "mutual_information_apc",
    n_positions: nPos,
    positions,
    matrix: apcMatrix,
    top_pairs: pairs.slice(0, topN),
  };
}
